src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,213 @@
+
+	@Override
+	public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry entry) {
+		String id = testIdentifier.getUniqueId();
+
+		// Read value from published entry
+		for (Map.Entry<String, String> singleEntry : entry.getKeyValuePairs().entrySet()) {
+			Property property = new Property(singleEntry.getKey(), singleEntry.getValue());
+			// Add to cache
+			if (propertyCache.containsKey(id)) {
+				propertyCache.get(id).add(property);
+			} else {
+				List<Property> currentProperties = new ArrayList<>();
+				currentProperties.add(property);
+				propertyCache.put(id, currentProperties);
+			}
+		}
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestcase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	Testcontainer processContainer(TestIdentifier container) {
+		Testcontainer testcontainer = new Testcontainer(container.getDisplayName());
+
+		// Add properties, if there are any
+		String uniqueId = container.getUniqueId();
+		Properties allProps = getProperties(uniqueId);
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcontainer.properties = allProps;
+		}
+
+		Set<TestIdentifier> allChildren = storedTestPlan.getChildren(uniqueId);
+
+		for (TestIdentifier identifier : allChildren) {
+			if (identifier.isContainer()) {
+				testcontainer.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				testcontainer.getTestcase().add(processTest(identifier));
+			}
+		}
+
+		return testcontainer;
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcase} for the report.
+	 *
+	 * @param test Object to be converted into {@code Testcase} element
+	 * @return Created object
+	 */
+	Testcase processTest(TestIdentifier test) {
+		String status = testStatusCache.getOrDefault(test.getUniqueId(), "UNKNOWN");
+		Testcase testcase = new Testcase(test.getDisplayName(), status);
+
+		// Add properties, if there are any
+		Properties allProps = getProperties(test.getUniqueId());
+
+		if (!allProps.getProperty().isEmpty()) {
```

Comment:
Is this check necessary? What would happen if you just set the properties without this check?

Reply:
If you take the alternative approach with `Optional` (see below), this could look like this:
```
if (allProps.isPresent()) {
    testcase.setProperties(allProps.get(());
}
```

Reply:
This looks a bit better to me because you don't break Law of Demeter with it?

Reply:
> Is this check necessary? What would happen if you just set the properties without this check?

Yes it is, because otherwise there would be a non null `Properties` element in the `Testcase` / `Testcontainer` class, which results in empty `<properties> elements for each object in the report, which I want to avoid.

```
<testcase name="an empty list, if no properties are stored" status="SUCCESSFUL">
    <properties/>
</testcase>

Reply:
> If you take the alternative approach with `Optional` (see below), this could look like this:

`Optional` is not serializeable. Even if , I would expect that it results in an empty `properties` element in the report.

Reply:
`Optional<Properties>` returned by `getProperties(String id)` does not leak into serialization because you unwrap it with `get()`.

I'm advocating for `Optional` because that would communicate clearly that `getProperties` (unlike a normal getter) might _or might not_ return with a value. It also makes the code a bit shorter.